pkg/objectio/writer.go

```diff
@@ -0,0 +1,174 @@
+		return err
+	}
+	reserved := make([]byte, HeaderReserved)
+	if err = binary.Write(&header, endian, reserved); err != nil {
+		return err
+	}
+	_, _, err = w.buffer.Write(header.Bytes())
+	return err
+}
+
+func (w *ObjectWriter) Write(batch *batch.Batch) (BlockObject, error) {
+	block := NewBlock(batch, w.object)
+	w.AddBlock(block.(*Block))
+	for i, vec := range batch.Vecs {
+		buf, err := vec.Show()
+		if err != nil {
+			return nil, err
+		}
+		offset, length, err := w.buffer.Write(buf)
+		if err != nil {
+			return nil, err
+		}
+		block.(*Block).columns[i].(*ColumnBlock).meta.location = Extent{
+			id:         block.GetMeta().header.blockId,
+			offset:     uint32(offset),
+			length:     uint32(length),
+			originSize: uint32(length),
+		}
+	}
+	return block, nil
+}
+
+func (w *ObjectWriter) WriteIndex(fd BlockObject, idx uint16, buf []byte) error {
+	var err error
+	block := w.GetBlock(fd.GetID())
+	if block == nil || block.columns[idx] == nil {
+		return errors.New("object io: not found")
+	}
+	offset, length, err := w.buffer.Write(buf)
+	if err != nil {
+		return err
+	}
+	block.columns[idx].(*ColumnBlock).meta.bloomFilter.offset = uint32(offset)
+	block.columns[idx].(*ColumnBlock).meta.bloomFilter.length = uint32(length)
+	block.columns[idx].(*ColumnBlock).meta.bloomFilter.originSize = uint32(length)
+	return err
+}
+
+func (w *ObjectWriter) WriteEnd() (map[uint32]BlockObject, error) {
+	var err error
+	w.RLock()
+	defer w.RUnlock()
+	var buf bytes.Buffer
+	for i, block := range w.blocks {
+		meta, err := block.(*Block).MarshalMeta()
+		if err != nil {
+			return nil, err
+		}
+		offset, length, err := w.buffer.Write(meta)
+		if err != nil {
+			return nil, err
+		}
+		w.blocks[i].(*Block).extent = Extent{
+			id:         block.(*Block).header.blockId,
+			offset:     uint32(offset),
+			length:     uint32(length),
+			originSize: uint32(length),
+		}
+		if err = binary.Write(&buf, endian, w.blocks[i].(*Block).extent.Offset()); err != nil {
+			return nil, err
+		}
+		if err = binary.Write(&buf, endian, w.blocks[i].(*Block).extent.Length()); err != nil {
+			return nil, err
+		}
+		if err = binary.Write(&buf, endian, w.blocks[i].(*Block).extent.OriginSize()); err != nil {
+			return nil, err
+		}
+	}
+	if err = binary.Write(&buf, endian, uint8(0)); err != nil {
+		return nil, err
+	}
+	if err = binary.Write(&buf, endian, uint32(len(w.blocks))); err != nil {
+		return nil, err
+	}
+	if err = binary.Write(&buf, endian, uint64(Magic)); err != nil {
+		return nil, err
+	}
+	_, _, err = w.buffer.Write(buf.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	return w.blocks, err
+}
+
+// Sync is for testing
+func (w *ObjectWriter) Sync(dir string) error {
+	err := w.object.fs.Write(context.Background(), w.buffer.GetData())
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (w *ObjectWriter) AddBlock(block *Block) {
+	w.Lock()
+	defer w.Unlock()
+	block.id = w.lastId
+	w.blocks[block.id] = block
+	w.lastId++
+}
+
```

Comment:
how to write meta data?

Reply:
maybe  user  only write bat.  
you can reset the meta and append to block?

Reply:
meta is generated by the writer according to the batch and written to the meta area.Execute write will return a "BlockObject" handle, use BlockObject.GetExtent() to get the location of meta in fileservice